// react-file-uploader/lib/UploadManager.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _invariant = require('invariant');

var _invariant2 = _interopRequireDefault(_invariant);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _bindKey = require('lodash/bindKey');

var _bindKey2 = _interopRequireDefault(_bindKey);

var _clone = require('lodash/clone');

var _clone2 = _interopRequireDefault(_clone);

var _debounce = require('lodash/debounce');

var _debounce2 = _interopRequireDefault(_debounce);

var _isEmpty = require('lodash/isEmpty');

var _isEmpty2 = _interopRequireDefault(_isEmpty);

var _superagent = require('superagent');

var _superagent2 = _interopRequireDefault(_superagent);

var _status = require('./constants/status');

var _status2 = _interopRequireDefault(_status);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var debug = require('debug')('react-file-uploader:UploadManager');

var UploadManager = function (_Component) {
  _inherits(UploadManager, _Component);

  function UploadManager(props) {
    _classCallCheck(this, UploadManager);

    var _this = _possibleConstructorReturn(this, (UploadManager.__proto__ || Object.getPrototypeOf(UploadManager)).call(this, props));

    _this.requests = {};
    _this.abort = _this.abort.bind(_this);
    _this.upload = _this.upload.bind(_this);
    _this.onProgress = (0, _debounce2.default)(_this.onProgress.bind(_this), props.progressDebounce);
    return _this;
  }

  _createClass(UploadManager, [{
    key: 'componentDidMount',
    value: function componentDidMount() {
      // eslint-disable-next-line react/prop-types
      if (this.props.uploadHeader) {
        // eslint-disable-next-line no-console
        console.warn('`props.uploadHeader` is DEPRECATED. Please use `props.uploadHeaderHandler` instead.');
      }

      // eslint-disable-next-line react/prop-types
      if (this.props.formDataParser) {
        // eslint-disable-next-line no-console
        console.warn('`props.formDataParser` is DEPRECATED. Please use `props.uploadDataHandler` instead.');
      }

      (0, _invariant2.default)(!!this.props.uploadUrl, 'Upload end point must be provided to upload files');

      (0, _invariant2.default)(!!this.props.onUploadEnd, 'onUploadEnd function must be provided');
    }
  }, {
    key: 'onProgress',
    value: function onProgress(fileId, progress) {
      var onUploadProgress = this.props.onUploadProgress,
          request = this.requests[fileId];


      if (request.xhr && request.xhr.readyState !== 4 && !request.aborted) {
        if (typeof onUploadProgress === 'function') {
          onUploadProgress(fileId, {
            progress: progress,
            status: _status2.default.UPLOADING
          });
        }
      }
    }
  }, {
    key: 'abort',
    value: function abort() {
      var file = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
      var onUploadAbort = this.props.onUploadAbort,
          request = this.requests[file.id];


      if (!request) {
        debug('request instance not found.');
        return;
      }

      request.abort();

      if (typeof onUploadAbort === 'function') {
        onUploadAbort(file.id, { status: _status2.default.ABORTED });
      }
    }
  }, {
    key: 'upload',
    value: function upload(url, file) {
      var _this2 = this;

      var _props = this.props,
          _props$reqConfigs = _props.reqConfigs,
          _props$reqConfigs$acc = _props$reqConfigs.accept,
          accept = _props$reqConfigs$acc === undefined ? 'application/json' : _props$reqConfigs$acc,
          _props$reqConfigs$met = _props$reqConfigs.method,
          method = _props$reqConfigs$met === undefined ? 'post' : _props$reqConfigs$met,
          timeout = _props$reqConfigs.timeout,
          _props$reqConfigs$wit = _props$reqConfigs.withCredentials,
          withCredentials = _props$reqConfigs$wit === undefined ? false : _props$reqConfigs$wit,
          onUploadStart = _props.onUploadStart,
          onUploadEnd = _props.onUploadEnd,
          uploadDataHandler = _props.uploadDataHandler,
          uploadErrorHandler = _props.uploadErrorHandler,
          uploadHeaderHandler = _props.uploadHeaderHandler;


      if (typeof onUploadStart === 'function') {
        onUploadStart(file.id, { status: _status2.default.UPLOADING });
      }

      var header = uploadHeaderHandler(file),
          data = uploadDataHandler(file);

      var request = _superagent2.default[method.toLowerCase()](url).accept(accept);

      if (!(0, _isEmpty2.default)(header)) {
        request.set(header);
      }

      if (timeout) {
        request.timeout(timeout);
      }

      if (withCredentials) {
        request.withCredentials();
      }

      this.requests[file.id] = request;

      debug('start uploading file#' + file.id + ' to ' + url, file);

      request.send(data).on('progress', function (_ref) {
        var percent = _ref.percent;
        return _this2.onProgress(file.id, percent);
      }).end(function (err, res) {
        var _uploadErrorHandler = uploadErrorHandler(err, res),
            error = _uploadErrorHandler.error,
            result = _uploadErrorHandler.result;

        if (error) {
          debug('failed to upload file', error);
        } else {
          debug('succeeded to upload file', result);
        }

        if (typeof onUploadEnd === 'function') {
          onUploadEnd(file.id, {
            progress: error && 0 || 100,
            error: error,
            result: error && undefined || result,
            status: error && _status2.default.FAILED || _status2.default.UPLOADED
          });
        }
      });
    }
  }, {
    key: 'render',
    value: function render() {
      var _this3 = this;

      var _props2 = this.props,
          component = _props2.component,
          customClass = _props2.customClass,
          style = _props2.style,
          children = _props2.children,
          uploadUrl = _props2.uploadUrl;


      return _react2.default.createElement(component, { className: (0, _classnames2.default)(customClass), style: style }, _react2.default.Children.map(children, function (child) {
        return (0, _react.cloneElement)(child, Object.assign({
          abort: (0, _bindKey2.default)(_this3, 'abort', child.props.file),
          upload: (0, _bindKey2.default)(_this3, 'upload', uploadUrl, child.props.file)
        }, child.props));
      }));
    }
  }]);

  return UploadManager;
}(_react.Component);

UploadManager.propTypes = {
  children: _propTypes2.default.oneOfType([_propTypes2.default.element, _propTypes2.default.arrayOf(_propTypes2.default.element)]),
  component: _propTypes2.default.string.isRequired,
  customClass: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.arrayOf(_propTypes2.default.string)]),
  onUploadAbort: _propTypes2.default.func,
  onUploadStart: _propTypes2.default.func,
  onUploadProgress: _propTypes2.default.func,
  onUploadEnd: _propTypes2.default.func.isRequired,
  progressDebounce: _propTypes2.default.number,
  reqConfigs: _propTypes2.default.shape({
    accept: _propTypes2.default.string,
    method: _propTypes2.default.string,
    timeout: _propTypes2.default.shape({
      response: _propTypes2.default.number,
      deadline: _propTypes2.default.number
    }),
    withCredentials: _propTypes2.default.bool
  }),
  style: _propTypes2.default.object,
  uploadDataHandler: _propTypes2.default.func,
  uploadErrorHandler: _propTypes2.default.func,
  uploadHeaderHandler: _propTypes2.default.func,
  uploadUrl: _propTypes2.default.string.isRequired
};

UploadManager.defaultProps = {
  component: 'ul',
  progressDebounce: 150,
  reqConfigs: {},
  uploadDataHandler: function uploadDataHandler(file) {
    var formData = new FormData();
    formData.append('file', file.data);
    return formData;
  },
  uploadErrorHandler: function uploadErrorHandler(err) {
    var res = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};

    var body = res.body ? (0, _clone2.default)(res.body) : {};
    var error = null;

    if (err) {
      error = err.message;
    } else if (body.errors) {
      error = body.errors;
    }

    delete body.errors;

    return { error: error, result: body };
  },
  uploadHeaderHandler: function uploadHeaderHandler() {
    return {};
  }
};

exports.default = UploadManager;
